### DISCUSSION TITLE
Run pkimetal as non-root

### DISCUSSION THREAD

**Discussion by dagleger2** [pkimetal/pkimetal#135] [Q&A] [answered]
Hi,
I have built a Docker image based on the pkimetal image published in the DHCR repo. The only addition I make is to create a non-root user and run pkimetal under this user instead of root. 

Under a non-root user, most linters work just fine - except for badkeys, where i get the following in the pkimetal response:

{
        "Linter": "pkimetal",
        "Finding": "badkeys: stdout.Scan() => false",
        "Severity": "fatal"
}

Is there any way to avoid this error and have badkeys return a normal response?
Our security policy requires containerized applications to run as a non-root user.

**Comment by robstradling** [comment]
The `./python3 badkeys-cli --update-bl` command in the Dockerfile downloads the badkeys blocklist and stores it at `~/.cache/badkeys/blocklist.dat`.  So when you run pkimetal as a different user, it's not finding that file.  You'll probably see this logged when you stop pkimetal:
```
{"level":"info","@timestamp":"2024-10-04T20:13:32.102Z","msg":"From stderr","instance#":0,"name":"badkeys","text":"blocklist metadata not found, you need to run --update-bl"}
```

I'll take a look at reworking the Dockerfile so that only the steps that require root privileges run as the root user.

**Reply by dagleger2** [reply]
Thanks Rob!

I should have noticed the log entry but missed it.

Now that I know what the issue is, I may be able to work around it.
From a security point of view, however, I think running applications as non-root is a good idea, even in containers.

Regards, 
Dag

**Comment by robstradling** [comment]
Hi @dagleger2.  Does https://github.com/pkimetal/pkimetal/pull/147 look OK to you?

**Reply by dagleger2** [reply]
Hi Rob,
This looks like exactly what I need.
Thanks!

**Reply by robstradling** [reply, marked as answer]
Great.  Thanks @dagleger2.  #147 merged.
